File: Aula_Ricardo/Programação_orientada_a_objetos/python/Classe_abstrata.py
from abc import ABC, abstractmethod

class Funcionario(ABC):
    @abstractmethod    
    def calcular_salario(self):
        pass
    

class FuncionarioHorista(Funcionario):
    def __init__(self, valor_hora, horas_trabalhadas):
        self.valor_hora = valor_hora
        self.horas_trabalhasdas = horas_trabalhadas

    def calcular_salario(self):
            return self.horas_trabalhasdas * self.valor_hora


class FuncionarioAssalariado(Funcionario):
    def __init__(self, salario_mensal):
        self.salario_mensal = salario_mensal

    def calcular_salario(self):
        return self.salario_mensal
    
horista = FuncionarioHorista(100, 20)
assalariado = FuncionarioAssalariado(2000)

print(horista.calcular_salario())
print(assalariado.calcular_salario())

File: Aula_Ricardo/Programação_orientada_a_objetos/go/Classe_abstrata.go
package main

import "fmt"

type Funcionario interface {
	CalcularSalario() float64
}

type FuncionarioHorista struct {
	ValorHora        float64
	HorasTrabalhadas float64
}

func (f FuncionarioHorista) CalcularSalario() float64 {
	return f.HorasTrabalhadas * f.ValorHora
}

type FuncionarioAssalariado struct {
	SalarioMensal float64
}

func (f FuncionarioAssalariado) CalcularSalario() float64 {
	return f.SalarioMensal
}

func main() {
	horista := FuncionarioHorista{ValorHora: 100, HorasTrabalhadas: 20}
	assalariado := FuncionarioAssalariado{SalarioMensal: 2000}

	fmt.Println(horista.CalcularSalario())
	fmt.Println(assalariado.CalcularSalario())
}
